src/backend/distributed/operations/shard_cleaner.c

```diff
@@ -216,6 +380,229 @@ DropOrphanedShards(bool waitForLocks)
+	 *    a) If isSuccess = true, drop cleanup records as operation is nearing completion.
+	 *       As the operation is nearing successful completion. This is done as part of the
+	 *       same transaction so will rollback in case of potential failure later.
+	 *
+	 *    b) If isSuccess = false, drop resource and cleanup records in a seperate transaction.
+	 */
+
+	/* We must have a valid OperationId. Any operation requring cleanup
+	 * will call StartNewOperationNeedingCleanup.
+	 */
+	Assert(CurrentOperationId != INVALID_OPERATION_ID);
+
+	List *currentOperationRecordList = ListCleanupRecordsForCurrentOperation();
+
+	int removedShardCountOnComplete = 0;
+	int failedShardCountOnComplete = 0;
+
+	CleanupRecord *record = NULL;
+	foreach_ptr(record, currentOperationRecordList)
+	{
+		/* We only supporting cleaning shards right now */
+		if (record->objectType != CLEANUP_OBJECT_SHARD_PLACEMENT)
+		{
+			ereport(WARNING, (errmsg("Invalid object type %d for cleanup record ",
+									 record->objectType)));
+			continue;
+		}
+
+		if (record->policy == CLEANUP_ALWAYS ||
+			(record->policy == CLEANUP_ON_FAILURE && !isSuccess))
+		{
+			char *qualifiedTableName = record->objectName;
+			WorkerNode *workerNode = LookupNodeForGroup(record->nodeGroupId);
+
+			if (TryDropShardOutsideTransaction(CurrentOperationId,
+											   qualifiedTableName,
+											   workerNode->workerName,
+											   workerNode->workerPort))
+			{
+				/*
+				 * Delete cleanup records outside transaction as:
+				 * 1.The resources are marked as 'CLEANUP_ALWAYS' and should be cleaned no matter
+				 *   you succeeded or failed.
+				 * 2.The resources are marked as 'CLEANUP_ON_FAILURE' and we know isSucess = FALSE
+				 *   i.e. operation is failing. Given we will abort the operation's transaction, we
+				 *   cannot delete records in the current transaction. We delete these records outside
+				 *   of the current transaction via a localhost connection.
+				 */
+				DeleteCleanupRecordByRecordIdOutsideTransaction(record->recordId);
+				removedShardCountOnComplete++;
+			}
+			else
+			{
+				failedShardCountOnComplete++;
+			}
+		}
+		else if (record->policy == CLEANUP_ON_FAILURE && isSuccess)
+		{
+			/* Delete cleanup records in same transaction as:
+			 * The resources are marked as 'CLEANUP_ON_FAILURE' and we are approaching a successful
+			 * completion of the operation. However, we cannot guarentee that operation will succeed
+			 * so we tie the Delete with parent transaction.
+			 */
+			DeleteCleanupRecordByRecordId(record->recordId);
+		}
+	}
+
+	if (list_length(currentOperationRecordList) > 0)
+	{
+		ereport(LOG, (errmsg("Removed %d orphaned shards out of %d",
+							 removedShardCountOnComplete, list_length(
+								 currentOperationRecordList))));
+
+		if (failedShardCountOnComplete > 0)
+		{
+			ereport(WARNING, (errmsg("Failed to drop %d cleanup shards out of %d",
```

Comment:
maybe: Failed to clean up %d out of %d shards

Reply:
Updated.